README.md
# Convert Windows DHCP to Cisco CLI

## Version 1.1.0

### Justin S. Cooksey - 2023-04-21

Script to export DHCP from a Windows Server DHCP database (exports to XML) and then convert the scope options and static leases to Cisco CLI to configure a router to take over the DHCP role on the network.

## Usage

```Powershell
Convert-WindowsDHCPToCisco
```

- It will ask for the DHCP host servername.
- It will require to have administrtor privilegde to be able to export the DHCP scopes.
- Working files are stored under the current %TEMP% path
- Output file is stored in the execution path.

## DHCP Options Documentation

[Wikipedia DHCP Options table](https://en.wikipedia.org/wiki/Dynamic_Host_Configuration_Protocol#Client_configuration_parameters)

[Actual RFC2312](https://tools.ietf.org/html/rfc2132)

## Currently handles DHCP Options

| Code | Option Description      | Cisco Output     |
| ---- | ----------------------- | ---------------- |
| 3    | Default Gateway         | default-router   |
| 4    | Time Server             | _ignoring_       |
| 6    | Domain Nameserver       | dns-server       |
| 15   | Domain Name             | domain-name      |
| 42   | NTP Servers             | option 42 ip     |
| 51   | Lease time              | _ignoring_       |
| 66   | TFTP Server             | next-server      |
| 67   | Boot filename           | bootfile         |
| 81   | MS DHCP Name Protection | _ignoring_       |
| 161  | FTP Server              | option 161 ip    |
| 162  | Path                    | option 162 ascii |
| 242  | Avaya                   | option 242 ascii |
| 252  | Proxy PAC URL           | option 252 ascii |


CHANGELOG.md
# Change Log

---

## Version 1.1.0

### Justin S. Cooksey - 2023-04-21

- Additrional option supporting 242

---

## Version 1.0.0

### Justin S. Cooksey - 2021-02-28

- Initial release


wip/convert121.ps1
<#
$array =@('0x10', '0x0A', '0x0A', '0xC0', '0xA8', '0xA9', '0xE7', '0x1C', '0xCB', '0x1C',
          '0x1C', '0x00', '0xC0', '0xA8', '0xA9', '0xF5', '0x18', '0xC0', '0xA8', '0x00',
          '0xC0', '0xA8', '0xA9', '0xE7')
#>
 
$array = @('0x10','0xc0','0xa8', '0xc0', '0xa8', '0x6e', '0x01','0x08', '0x0a', '0xc0', '0xa8', '0xa8', '0x05')

Write-Host $array

$strNetwork = ""
$atMask = $true

for($i=0; $i -lt $array.Length; $i++) {

    if($atMask) {
        $mask = [convert]::toint64($array[$i],16)
        
        if($mask -lt 9) { 
            $strNetwork = $strNetwork + [convert]::toint64($array[$i+1],16) + ".0.0.0"
            $i = $i + 1
        }
        elseif($mask -lt 17) { 
            $strNetwork = $strNetwork + [convert]::toint64($array[$i+1],16)
            $strNetwork = $strNetwork + "." + [convert]::toint64($array[$i+2],16) + ".0.0"
            $i = $1 + 2
        }
        elseif($mask -lt 25)  { 
            $strNetwork = $strNetwork + [convert]::toint64($array[$i+1],16)
            $strNetwork = $strNetwork + "." + [convert]::toint64($array[$i+2],16)
            $strNetwork = $strNetwork + "." + [convert]::toint64($array[$i+3],16) + ".0"
            $i = $i + 3
        }
        elseif($mask -ge 25)  { 
            $strNetwork = $strNetwork + [convert]::toint64($array[$i+1],16)
            $strNetwork = $strNetwork + "." + [convert]::toint64($array[$i+2],16)
            $strNetwork = $strNetwork + "." + [convert]::toint64($array[$i+3],16)
            $strNetwork = $strNetwork + "." + [convert]::toint64($array[$i+4],16)
            $i = $i + 4
        }
        $strNetwork = $strNetwork + "/" + $mask
        
        $strRouter = ""
        $strRouter = $strRouter + [convert]::toint64($array[$i+1],16)
        $strRouter = $strRouter + "." + [convert]::toint64($array[$i+2],16)
        $strRouter = $strRouter + "." + [convert]::toint64($array[$i+3],16)
        $strRouter = $strRouter + "." + [convert]::toint64($array[$i+4],16)

        $i = $i + 4

        Write-Host($strNetwork + " " + $strRouter)

        $mask = $true
        $strNetwork = ""

    }
    
}

#
# Mask is
#   m < 9   (8)     1 octet 
#   m < 17  (16)    2 octets
#   m < 25  (24)    3 octets
#                   4 octets

#
# option 121 hex 10c0a8c0a86e01 080AC0A8A805
# Written from Cisco as
#  option 121 hex 10c0.a8c0.a86e.0108.0ac0.a8a8.05


Convert-WindowsDHCPToCisco.ps1
#
# Convert-WindowsDHCPToCisco
#
# Justin S Cooksey https://github.com/jscooksey/Convert-WindowsDHCPToCisco
#
# Release 2023-04-21
#

#                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
# Convert an exclusion range from MSDHCP XML to Cisco CLI code
#

function Convert-Exclusion {
    param( $Range )

    return("ip dhcp excluded-address " + $Range.StartRange + " " + $Range.EndRange)
}

#
# Convert an DHCP Pool from MSDHCP XML to Cisco CLI code
#

function Convert-PoolName {
    param( [string]$poolName )

    $result = $poolName.Split('.')[0]
    $result = [string]$result.Replace(" ", "-")
    $result = [string]$result.ToLower()

    return $result
}

#
# Convert an lease duration from MSDHCP XML to Cisco CLI code
#

function Convert-LeaseDuration {
    param( [string]$leaseDuration)

    $textReformat = $leaseDuration.Replace( ",", ".")
    $seconds = ([TimeSpan]::Parse($textReformat)).TotalSeconds
    $days = [int]($seconds / 86400)
    $hours = [int]($seconds / 3600) % 24
    $minutes = [int]($seconds / 60) % 60
    $result = "lease " + [string]$days + " " + [string]$hours + " " + [string]$minutes
    return  $result
}

#
# Convert DHCP options from MSDHCP XML to Cisco CLI code
#

function Convert-DHCPOptions {
    param( $Options)

    $result = ""

    foreach ($option in $Options) {
        switch ($option.OptionId) {
            3 { $result = $result + "  default-router " + $option.Value + "`n"; break }
            4 { break }                                                           #Time Server, dont use
            6 { $result = $result + "  dns-server " + $option.Value + "`n"; break }
            15 { $result = $result + "  domain-name " + $option.Value + "`n"; break }
            42 { $result = $result + "  option 42 ip " + $option.Value + "`n"; break } # NTP Servers
            51 { break } #Lease time in seconds. Not Required
            66 { $result = $result + "  next-server " + $option.Value + "`n"; break } #DAP TFTP Server to load boot config
            67 { $result = $result + "  bootfile " + $option.Value + "`n"; break } #DAP Bootfile name see option 66                                          
            81 { break } #Client FQDN. Not Required
            161 { $result = $result + "  option 161 ip " + $option.Value + "`n"; break }
            162 { $result = $result + '  option 162 ascii "' + $option.Value + '"' + "`n"; break }
            242 { $result = $result + '  option 242 ascii "' + $option.Value + '"' + "`n"; break }
            252 { $result = $result + '  option 252 ascii "' + $option.Value + '"' + "`n"; break }
            default { $result = $result + "*** Unknown Option " + $option.OptionId + " with Value of " + $option.Value + " ***" + "`n" }
        }
    }

    return  $result
}

#
# Convert a MAC Address from MSDHCP XML to Cisco CLI code
#

function Convert-MAC {
    param( [string]$msMAC)

    $result = ""

    $result = "01" + $msMac.Replace("-", "")
    $result = $result.Insert(4, '.')
    $result = $result.Insert(9, '.')
    $result = $result.Insert(14, '.')
    return $result
}

#
# Convert DHCP reservations from MSDHCP XML to Cisco CLI code
#

function Convert-Reservations {
    param( $Reservations, [string]$subnetMask, [string]$subnetOptions)

    $result = ""

    foreach ($reservation in $Reservations) {
        $result = $result + "ip dhcp pool " + (Convert-PoolName -poolName $reservation.Name) + "`n"
        $result = $result + "  host " + $reservation.IPAddress + " " + $subnetMask + "`n"
        $result = $result + "  client-identifier " + (Convert-MAC -msMAC $reservation.ClientId) + "`n"
        $result = $result + $subnetOptions + "`n"
    }

    return  $result
}


#
# Primary code to read XML file and output Cisco CLI code
#

$OldServer = Read-Host 'Enter the computer name of the DHCP server to export from: '
$xmlFile = $env:TEMP + "\DHCPExport.xml"
Export-DhcpServer -ComputerName $OldServer -File $xmlFile

[XML]$fullDHCP = Get-Content $xmlFile
$allScopes = $fullDHCP.DHCPServer.IPv4.Scopes.Scope

$globalOptions = Convert-DHCPOptions -Options $fullDHCP.DHCPServer.IPv4.OptionValues.OptionValue

foreach ($scope in $allScopes) {
    foreach ($exclusionRange in $scope.ExclusionRanges.IPRange) {
        $result = Convert-Exclusion -Range $exclusionRange
        Write-Output -InputObject $result
    }
    Write-Output -InputObject $null

    Write-Output -InputObject ("ip dhcp pool " + (Convert-PoolName -poolName $scope.Name))
    Write-Output -InputObject ("  network " + $scope.ScopeId + " " + $scope.SubnetMask)
    Write-Output -InputObject ("  " + (Convert-LeaseDuration -leaseDuration $scope.LeaseDuration))
    Write-Output -InputObject ("  update dns both override")
    $subnetOptions = $globalOptions + (Convert-DHCPOptions -Options $scope.OptionValues.OptionValue)
    Write-Output -InputObject ($subnetOptions)

    Write-Output -InputObject (Convert-Reservations -Reservations $scope.Reservations.Reservation -subnetMask $scope.SubnetMask -subnetOptions $null)

}
